proposals/NNNN-isolated-default-values.md

```diff
@@ -0,0 +1,173 @@
+
+## Detailed design
+
+### Inference of default value isolation requirements
+
+Default value expressions are always evaluated synchronously. All calls that are made during the evaluation of the expression must also be synchronous. If the callee is isolated, then the default value expression must already be in the same isolation domain in order to make the call synchronously. So, for a given default value expression, the inferred isolation is the required isolation of its subexpressions. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+
+@MainActor func useDefault(value: Int = requiresMainActor()) { ... }
+```
+
+In the above code, the default argument for `value` requires `@MainActor` isolation, because the default value calls `requiresMainActor()` which is isolated to `@MainActor`.
+
+A default value expression must only have one required isolation; it is an error for a default value expression to contain multiple callees with different actor isolation. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+@AnotherActor func requiresAnotherActor() -> Int { ... }
+
+@MainActor func useDefault(
+  value: (Int, Int) = (requiresMainActor(), requiresAnotherActor()) // error!
+) {}
+```
+
+The above example is invalid because the default argument for `value` requires both `@MainActor` and `@AnotherActor`, but the caller can never satisfy both isolation requirements simultaneously.
+
+#### Closures
+
+Evaluating a closure literal itself can happen in any isolation domain; the actor isolation of a closure only applies when calling the closure. An actor-isolated closure enables the closure body to make calls within that isolation domain synchronously. For a closure literal in a default value expression that is not explicitly annotated with actor isolation, the inferred isolation of the closure is the union of isolation contexts of all callees in the closure body for synchronous calls. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+
+@MainActor func useDefaultClosure(
+  closure: () -> Void = {
+    requiresMainActor()
+  }
+) {}
+```
+
+The above `useDefaultClosure` function has a default argument value that is a closure literal. The closure body calls a `@MainActor`-isolated function synchronously, therefore the closure itself must be `@MainActor` isolated.
+
+Note that the inferred actor isolation of a closure literal can never be an actor instance based on the following two properties:
```

Comment:
I tripped over this sentence a bit, because I interpreted it as "you can't have a closure literal with actor instance isolation in a default argument". You can, if it's written as something like `{ (isolated a: MyActor) in ... }`. What this means to say is that the inference algorithm will never infer such a thing.

Reply:
Oops, yeah the "inferred" in this sentence is load bearing and easy to miss. I'll make this point more clear.